hoodie-common/src/main/java/com/uber/hoodie/common/HoodieFileGroupBuilder.java

```diff
@@ -0,0 +1,184 @@
+  /**
+   * Partition containing the file group.
+   */
+  private final String partitionPath;
+  /**
+   * uniquely identifies the file group
+   */
+  private final String id;
+  /**
+   * Timeline, based on which all getter work
+   */
+  private final HoodieTimeline timeline;
+
+  public HoodieFileGroupBuilder(String partitionPath, String id, HoodieTimeline timeline) {
+    this.partitionPath = partitionPath;
+    this.id = id;
+    this.timeline = timeline;
+  }
+
+  /**
+   * Add a new datafile into the file group
+   */
+  public HoodieFileGroupBuilder withDataFile(HoodieDataFile dataFile) {
+    // skip the data file if it is not from a committed instant. Concurrent
+    // compaction could be running and there could be partial/complete data-files
+    // without the compaction instant marked complete. We should skip these till the compaction
+    // instant is marked committed.
+    if (timeline.containsOrBeforeTimelineStarts(dataFile.getCommitTime())) {
+      if (!fileSlices.containsKey(dataFile.getCommitTime())) {
+        fileSlices.put(dataFile.getCommitTime(), new FileSlice(dataFile.getCommitTime(), id));
+      }
+      fileSlices.get(dataFile.getCommitTime()).setDataFile(dataFile);
+    } else {
+      log.warn("Skipping data file " + dataFile + " when building file-group as its instant time is not valid");
+    }
+    return this;
+  }
+
+  /**
+   * Add a new log file into the group
+   */
+  public HoodieFileGroupBuilder withLogFile(HoodieLogFile logFile) {
+    // We allow log-files whose base-commit is that of incomplete compaction instant to be added to the file-group.
+    // It is possible that the log-file was generated because of an inflight delta-instant. Even in this case, the log
+    // file will be allowed to added to the file-groups. The log-blocks(log-files) belonging to inflight delta-instants
+    // are filtered-out during record-reading.
+    if (!fileSlices.containsKey(logFile.getBaseCommitTime())) {
+      fileSlices.put(logFile.getBaseCommitTime(), new FileSlice(logFile.getBaseCommitTime(), id));
+    }
+    fileSlices.get(logFile.getBaseCommitTime()).addLogFile(logFile);
+    return this;
+  }
+
+  @Override
+  public HoodieFileGroup build() {
+    Preconditions.checkNotNull(id, "File Id must not be null");
+    Preconditions.checkNotNull(timeline, "Timeline must not be null");
+
+    // We can have at-most last 2 file-slices without base file. This is possible if a compaction request
+    // has been set for a file-group first time. In this case, the file group may not even have a data file
+    // (in the case of log files supporting inserts).
+    Preconditions.checkArgument(fileSlices.values().stream().filter(f -> !f.getDataFile().isPresent()).count() <= 2);
+    Map<String, FileSlice> mergedFileSlices = fileSlices;
+    if (!fileSlices.isEmpty()) {
+      FileSlice lastSlice = fileSlices.lastEntry().getValue();
```

Comment:
Similar concerns as above, should we do this in 2 different layers, basically one layer provides a vanilla group of all file groups available and then the second layer filters on them ? The data from inner layer could be made unavailable to outer packages ?

Reply:
At file-slice level,  I am hesitant to lazily merge and create correct file-slices because :
(a) Merging does not result in loss of information - A merged file-slice retains all the information to track underlying physical files. 
(b) Early merging keeps the definition of file-slice consistent  - A file-slice gives a complete snapshot view of the file-group.  With lazy file-slice merging, a file-group view is spread across 1 or 2  file-slices

But, we can chat about it on Monday to resolve this.